blog/2024-11-22-inspektor-gadget-unittesting-framework.md

```diff
@@ -0,0 +1,265 @@
+gadgetRunner.RunGadget()
+```
+
+### Validating captured output against the expected output 
+
+For every particular testCase a specific validation function can be defined in similar fashion:
+
+```go
+validateEvent: func(t *testing.T, info *utilstest.RunnerInfo, fd int, events []ExpectedTraceOpenEvent) {
+        utilstest.ExpectOneEvent(func(info *utilstest.RunnerInfo, fd int) *ExpectedTraceOpenEvent {
+            return &ExpectedTraceOpenEvent{
+                Proc:  info.Proc,
+                Fd:    uint32(fd),
+                FName: "/dev/null",
+            }
+        })(t, info, fd, events)
+}
+```
+
+In the given example, the test expects only one event to be captured. Similarly, other functions like `ExpectNoEvent` and `ExpectAtLeastOneEvent` can be used. For checking the equality of two fields, `require.Equal` can be utilized along with other supported functions from the require package.
+
+### Selecting the Testing Environment:
+
+You can configure the gadget repository and tag, as well as the image verification method, by setting the following environment variables:
+
+- **GADGET_REPOSITORY:**
+Specifies the repository containing the Gadget image. Default is ghcr.io/inspektor-gadget/gadget
+
+- **GADGET_TAG:**
+Specifies the tag for the Gadget image. 
+
+
+- **IG_VERIFY_IMAGE:**
+Determines weather to do the image verification or not. Incase of testing gadget on different kernel version this should be kept false since we use [vimto](https://github.com/lmb/vimto) as vimto doesn’t support networking capabilities.
+
+### Running the Test
+On normal testing machine we can use straight forward commands.
+```go
+# [Optional] If running the test for a gadget whose image resides in a remote container registry
+export GADGET_REPOSITORY=ghcr.io/my-org GADGET_TAG=latest
+
+# Run the test
+go test -exec 'sudo -E' -v
+```
+
+### Testing on Different Kernel Versions
+For running unit tests for different gadgets on various kernel versions, you need to use vimto.
+
+First, set up QEMU by installing the required packages (qemu-system-x86) and updating the necessary permissions for /dev/kvm.
+
+Next, install vimto using the go install command with the latest version available from its repository.
+
+```bash
+# Install QEMU
+sudo apt-get update
+sudo apt-get install -y qemu-system-x86
+sudo chmod 666 /dev/kvm
+
+# Install vimto
+CGO_ENABLED=0 GOBIN=$(go env GOPATH)/bin go install lmb.io/vimto@latest
+
+# Running the Test
+export IG_VERIFY_IMAGE=false
+export VIMTO_VM_MEMORY=4906M
+export KERNEL_REPOSITORY=ghcr.io/mauriciovasquezbernal/ci-kernels
+export KERNEL_VERSION=5.10
+
+sudo -E vimto -kernel $(KERNEL_REPOSITORY):$(KERNEL_VERSION) -memory $(VIMTO_VM_MEMORY) -- go test -v .
+```
+Note: Since vimto doesn't support networking out side the Virtual Machine made the gadget image needs to pulled sepreately and present locally, we also need to keep the gadget image verification to false using appropiate flag due the no network capability available.
```

Comment:
```suggestion
Note: Since vimto doesn't support networking out side the Virtual Machine made the gadget image needs to pulled separately and present locally, we also need to keep the gadget image verification to false using the appropriate flag.
```